File: Problems/Spread_Spectrum_Radar.py

__author__ = "Ivón Olarte Rodríguez"

# Based on the work made by Kudela, Stripinis and Paulavicius


### Importing libraries
import numpy as np

### ===============================
### Spread Spectrum Radar 
### ================================



def Spread_Spectrum_Radar(x:np.ndarray) -> float:
    r"""
    Spread Spectrum Radar function.
    This function is a MATLAB translation of the original code provided.
    """
    assert x.ndim < 2, "Input array must be less than two-dimensional."

    n = x.size
    lb = get_xl(n)
    ub = get_xu(n)
    x_eval = np.abs(ub - lb)*x + lb
    
    # Determine the number of variables and initialize variables
    var = 2 * n - 1 
    hsum = np.zeros((2*var, ))
    
    # Calculate the values of hsum vector
    for kk in range(2 * var):
        if kk % 2 != 0:
            # Odd index
            i = int((kk + 1) / 2)
            hsum[kk] = 0
            for j in range(i,n):
                summ = 0
                for i1 in range(abs(2 * i - j - 1) ,j+1):
                    summ += x_eval[i1] 
                hsum[kk] += np.cos(summ) 

            hsum[kk] +=  0.5
        else:
            # Even index
            i = int(kk / 2)
            hsum[kk] = 0
            for j in range(i,n):
                summ = 0
                for i1 in range((abs(2 * i - j)),j+1):
                    summ += x_eval[i1] 
                
                hsum[kk] += np.cos(summ) 
            
            
        
    
    
    # Calculate the maximum value of hsum as the fitness value y
    y = max(hsum)
    
    # Check for NaN value and set a large value if necessary
    if np.isnan(y):
        y = 10**100
    
    return y


def get_xl(nx:int)->np.ndarray:
    r"""Lower bounds for the design variables."""
    xl = np.zeros((nx, ))
    return xl

def get_xu(nx:int)->np.ndarray:
    r"""Upper bounds for the design variables."""
    xu = np.ones((nx, ))*2*np.pi

    return xu


if __name__ == "__main__":
    # Test the function with a sample input
    # x = np.asarray([3.14159265358980, 3.14159265358980, 3.14159265358980, 3.14159265358980,	3.14159265358980,
    #                 	3.14159265358980,	3.14159265358980,	3.14159265358980,	3.14159265358980,	3.14159265358980,
    #                         	3.14159265358980,	3.14159265358980,	3.14159265358980,	3.14159265358980,	3.14159265358980,
    #                                 	3.14159265358980,	3.14159265358980,	3.14159265358980,	3.14159265358980,	3.14159265358980]).ravel()
    
    x = np.asarray([1.04719755119660, 3.14159265358980,	3.14159265358980,	3.14159265358980,	3.14159265358980,
                    	3.14159265358980,	3.14159265358980,	3.14159265358980,	3.14159265358980,	3.14159265358980,
                            	3.14159265358980,	3.14159265358980,	3.14159265358980,	3.14159265358980,	3.14159265358980,
                                    	3.14159265358980, 3.14159265358980, 3.14159265358980, 3.14159265358980, 3.14159265358980]).ravel()
    potential = Spread_Spectrum_Radar(x)
    print(f"Spread Spectrum Radar: {potential}")

File: Problems/Spread_Spectrum_Radar.m
function y = Spread_Spectrum_Radar(x)
    if size(x, 1) > size(x, 2), x = x'; end
    n = length(x);
    lb = get_xl(n)';
    ub = get_xu(n)';
    x = abs(ub - lb).*x + lb;
    
    % Determine the number of variables and initialize variables
    [~, d] = size(x);
    var = 2 * d - 1;
    hsum = zeros(var, 1);
    
    % Calculate the values of hsum vector
    for kk = 1:2 * var
        if rem(kk, 2) ~= 0
            % Odd index
            i = (kk + 1) / 2;
            hsum(kk) = 0;
            for j = i:d
                summ = 0;
                for i1 = (abs(2 * i - j - 1) + 1):j
                    summ = x(i1) + summ;
                end
                hsum(kk) = cos(summ) + hsum(kk);
            end
        else
            % Even index
            i = kk / 2;
            hsum(kk) = 0;
            for j = (i + 1):d
                summ = 0;
                for i1 = (abs(2 * i - j) + 1):j
                    summ = x(i1) + summ;
                end
                hsum(kk) = cos(summ) + hsum(kk);
            end
            hsum(kk) = hsum(kk) + 0.5;
        end
    end
    
    % Calculate the maximum value of hsum as the fitness value y
    y = max(hsum);
    
    % Check for NaN value and set a large value if necessary
    if isnan(y)
        y = 10^100;
    end
end

function xl = get_xl(nx)
    xl = zeros(nx, 1);
end

function xu = get_xu(nx)
    xu = ones(nx, 1).*2*pi;
end